eclipse/sw360
# feat(ui) : Strengthen sw360 admin privileges about Read and Write

[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.
> * Which issue is this pull request belonging to and how is it solving it? (*Refer to issue here*)
> * Did you add or update any new dependencies that are required for your change?

Issue:  #1290 #1412 #1450

# The outline of a functional improvement

## Purpose
This modification allows Admin user to access projects set by other users as PRIVATE or ME AND MODERATORS. This function can be enabled or disabled in the properties. (Default setting is disabled)

## The outline of function change
- Admin user can list all projects regardless of the project's visibility setting.  
    Figure1 List including projects set by other users as PRIVATE or ME AND MODERATORS  
   ![image](https://user-images.githubusercontent.com/39894802/154875970-a6f7c005-4a06-4c6e-a299-67272d621c2a.png)
- Admin user can view all projects regardless of the project's visibility settings.  
    - Admin user can view project details set by other users as PRIVATE or ME AND MODERATORS.  
    Figure2 Project detail view.  
   ![image](https://user-images.githubusercontent.com/39894802/154875993-69bf92c9-d5ad-4401-a074-b7708126fec2.png)
    - Admin user can view project details linked to other users' PRIVATE or ME AND MODERATORS projects.
      
- Whether to enable the above functions can be changed by the configuration file.  
    This function is enabled only when the following values ​​are added to sw360.properties.  
    admin.private.project.access.enabled=true  
    （If the value does not exist or false is specified, this function is disabled.）  

## Points of code change
- Backend changes  
    - The acquisition process of the accessible projects  
        - Source code  
            - backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectRepository.java  
                - getAccessibleProjectsSummary method  
                - getMyAccessibleProjectsCount method
        - Changes  
            Admin user can get all the project data from couchdb regardless of the  visibility settings.
    - ProjectPermissions  
        - Source code    
            - libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/permissions/ProjectPermissions.java  
                - isVisible method  
        - Changes  
            Admin user has READ permission for projects created by other users regardless of the visibility.  

    - Comparison of behavior when the function is enabled or disabled.  
        |Action|Private Access - Disable|Private Access - Enable|
        |-|-|-|
        |READ (other user PRIVATE project)|-|x|
        |READ (other user ME_AND_MODERATORS project )|-|x|
        |READ (other user BUISNESSUNIT_AND_MODERATORS project)|x|x|
        |READ (other user EVERYONE project)|x|x|

        x : Access allowed  
        \- : Access denied  
    
## How to do test
- Advance Preparations  
    1. Sign in as admin@sw360.org user and create the following projects.  
        - MyProject001 (Visibility:Everyone)
        - MyProject002 (Visibility:Everyone)
    1. Sign in as admin1@sw360.org user and create the following projects.  
        - OtherUserProject001 (Visibility:Private)
        - OtherUserProject002 (Visibility:Me and Moderators)
    1. Link each OtherUserProject as a subproject of MyProject.  
        - MyProject001 -> OtherUserProject001
        - MyProject002 -> OtherUserProject002
- Operation check  
    - Enable the function and check the following operation. (admin.private.project.access.enabled=true)  
        - Check that OtherUserProject001-002 are displayed on the list of Project section.  
        - Open any project and click the "Link to Projects" button. Check that OhterUserProject001-002 are displayed on the list.  
        - Click OtherUserProject001-002 on the list of Project section and check that the project detail view can be opened.  
        - Check that OtherUserProject001-002 can be opened on the edit view and the edited contents are reflected on the project detail view and Change log.  
        - Click MyProject001-002 on the list of Project section (these are linked to other users' PRIVATE or ME AND MODERATORS projects) and check that the project detail view can be opened.  
    - Disable the function and check the following operation. (admin.private.project.access.enabled=false)  
        - Check that OtherUserProject001-002 are not displayed on the list of Project section.   
        - Open any project and click the "Link to Projects" button. Check that OtherUserProject001-002 are not displayed on the list.  
        - Check that the URL of the project detail view of OtherUserProject001-002 are restricted.   
        - Check that the URL of the project edit view of OtherUserProject001-002 are restricted.          
        - Click MyProject001-002 on the list of Project section (these are linked to other users' PRIVATE or ME AND MODERATORS projects) and check that the project detail view is restricted.  
- Automatic test  
    Execute RunPrivateProjectAccessTest as a visibility test.  
    RunPrivateProjectAccessTest is changed so that it can be tested by switching the enable / disable of the function by specifying the system property. (-DRunPrivateProjectAccessTest=true or false)  
    Check that the automated test passes in each case.
